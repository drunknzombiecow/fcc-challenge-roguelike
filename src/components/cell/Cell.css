.Cell {
    position: absolute;
    box-sizing: border-box;
    background-size: contain;
    background-image: url("../../resources/img/textures/gravel.jpg");
}

.Cell::before {
    position: absolute;
    z-index: 1;
    left: 50%;
    top: 50%;
    width: 1px;
    height: 1px;
    border-radius: 50%;
    content: '';
}

.Cell--rock {
    z-index: 3;
    background-color: #666;
    background-image: url("../../resources/img/textures/blackbrick.jpg");
}

.Cell__content {
    left: 0;
    top: 0;
    width: 100%;
    height: 100%;
    background: transparent none center center no-repeat;
    background-size: contain;
    position: absolute;
    z-index: 2;
}

.Cell--player .Cell__content {
    background-image: url("../../resources/img/hero.svg");
}

.Cell--player::before {
    box-shadow: 0 0 20px 20px rgba(255, 255, 255, 0.5);
}

.Cell--health .Cell__content {
    background-image: url("../../resources/img/potion.svg");
}

.Cell--health::before {
    box-shadow: 0 0 20px 20px rgba(50, 205, 50, 0.5);
}

.Cell--weapon .Cell__content {
    background-image: url("../../resources/img/sword.svg");
}

.Cell--weapon::before {
    box-shadow: 0 0 20px 20px rgba(238, 232, 170, 0.5);
}

.Cell--enemy::before {
    box-shadow: 0 0 20px 20px rgba(240, 128, 128, 0.5);
}

.Cell--boss::before {
    box-shadow: 0 0 20px 20px rgba(255, 0, 0, 0.75);
}

.Cell--exit .Cell__content {
    background-image: url("../../resources/img/stairs.svg");
}

.Cell--exit::before {
    box-shadow: 0 0 20px 20px rgba(238, 130, 238, 0.75);
}

.Cell--exit {
    background-color: purple;
}

.Cell--outOfSight {
    display: none;
}