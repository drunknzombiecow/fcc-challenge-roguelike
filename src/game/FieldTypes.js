import {switchKeysAndValues} from '../utils/Object';

const Types = Object.freeze({
    rock: 0,
    earth: 1,
    wall: 2,
    player: 3,
    enemy: 4,
    boss: 5,
    exit: 6,
    weapon: 7,
    health: 8
});

const TypesByCode = Object.freeze(switchKeysAndValues(Types));

export default { Types, TypesByCode };
