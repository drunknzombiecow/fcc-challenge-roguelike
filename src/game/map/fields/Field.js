// @flow
import uuid from 'uuid';

import type { FieldTypeCode } from './FieldTypes';
import type { Point } from '../../../types/BasicTypes';

type FieldProperties = {
    x: number,
    y: number,
    type: FieldTypeCode,
    image?: string,
    id?: string
};

export default class Field {
    getPosition: Function;
    getType: Function;
    setType: Function;
    getImage: Function;
    setImage: Function;
    getId: Function;
    getState: Function;
    constructor(params: FieldProperties) {
        const { x, y, type, image } = params;
        const id = params.id || uuid.v4();

        function getPosition(): Point {
            return Object.freeze({ x, y });
        }

        function getType(): FieldTypeCode {
            return type;
        }

        function setType(newType: FieldTypeCode): Field {
            return new Field({ x, y, image, id, type: newType });
        }

        function getId(): string {
            return id;
        }

        function getImage(): ?string {
            return image;
        }

        function setImage(newImage: string): Field {
            return new Field({ x, y, type, id, image: newImage });
        }

        function getState(): FieldProperties {
            return Object.freeze({ x, y, type, id, image });
        }

        this.getPosition = getPosition;
        this.getId = getId;
        this.getState = getState;
        this.getType = getType;
        this.setType = setType;
        this.getImage = getImage;
        this.setImage = setImage;

        Object.freeze(this);
    }
}
