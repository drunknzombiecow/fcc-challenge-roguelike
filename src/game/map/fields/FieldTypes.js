// @flow
import { switchKeysAndValues } from '../../../utils/ObjectUtils';

export const Types: any = {
    rock: 0,
    earth: 1,
    wall: 2,
    player: 3,
    enemy: 4,
    boss: 5,
    exit: 6,
    weapon: 7,
    health: 8
};

export const TypesByCode = Object.freeze(switchKeysAndValues(Types));

export type FieldTypeCode = $Keys<TypesByCode>;
export type FieldTypeName = $Keys<Types>;

export default { Types, TypesByCode };
