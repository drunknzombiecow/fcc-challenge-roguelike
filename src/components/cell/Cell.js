import React from 'react';
import FieldTypes from '../../game/FieldTypes';
import './Cell.css';

export function Cell(props) {
    props.game.dispatch('cellHasType', { x: props.x, y: props.y, type: props.type });
    return (
        <div className={'Cell Cell--' + FieldTypes.TypesByCode[props.type]}></div>
    );
}

export default Cell;
