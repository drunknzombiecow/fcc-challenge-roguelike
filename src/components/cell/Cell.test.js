import React from 'react';
import renderer from 'react-test-renderer';
import Cell from './Cell';
import Field from '../../game/map/fields/Field';
import { Types } from '../../game/map/fields/FieldTypes';

function createNewCell(props) {
    const {
        field,
        cellSize = 0,
        cellPaddingTop = 0,
        cellPaddingLeft = 0,
        playerPosition = {
            x: 15,
            y: 15
        }} = props;
    return renderer.create(<Cell
        field={field}
        key={field.getId()}
        cellSize={cellSize}
        cellPaddingTop={cellPaddingTop}
        cellPaddingLeft={cellPaddingLeft}
        playerPosition={playerPosition}
    />);
}

test('Cell is not rendered when it is out of sight', () => {
    const field = new Field({
        x: 1,
        y: 1,
        type: Types.earth
    });

    const cell = createNewCell({ field });
    const cellAsTree = cell.toJSON();
    expect(cellAsTree).toMatchSnapshot();
});

test('Cell is rendered when it is in sight', () => {
    const field = new Field({
        x: 14,
        y: 14,
        type: Types.earth
    });

    const cell = createNewCell({ field });
    const cellAsTree = cell.toJSON();
    expect(cellAsTree).toMatchSnapshot();
});