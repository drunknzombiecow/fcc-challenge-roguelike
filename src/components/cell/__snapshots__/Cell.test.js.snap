// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell is not rendered when it is out of sight 1`] = `null`;

exports[`Cell is rendered ehwn it is in sight 1`] = `
<div
  className="Cell Cell--earth"
  style={
    Object {
      "height": 0,
      "left": 0,
      "top": 0,
      "width": 0,
    }
  }
>
  <div
    className="Cell__content"
    style={undefined}
  />
</div>
`;
